tests/integration/application-charm/src/charm.py

```diff
@@ -0,0 +1,198 @@
+        self.framework.observe(self.on[REMOTE].relation_broken, self._on_database_broken)
+
+    def _on_start(self, _) -> None:
+        """Only sets an waiting status."""
+        self.unit.status = WaitingStatus("Waiting for relation")
+
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info("Received database data")
+
+        if not self.unit.is_leader():
+            return
+
+        # connection configuration
+        config = {
+            "user": event.username,
+            "password": event.password,
+            "host": event.endpoints.split(":")[0],
+            "database": self.database_name,
+            "raise_on_warnings": False,
+        }
+
+        with MysqlConnector(config) as cursor:
+            self._create_test_table(cursor)
+
+            self._insert_test_data(
+                cursor,
+                event.username,
+                event.password,
+                event.endpoints,
+                event.version,
+                event.read_only_endpoints,
+            )
+
+        self._peers.data[self.app]["inserted"] = "true"
+        logger.info("Inserted relation data in database")
+        self.unit.status = ActiveStatus()
+
+    def _on_database_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"first database endpoints have been changed to: {event.endpoints}")
+
+    def _on_peer_relation_changed(self, event: RelationChangedEvent):
+        """Handle peer relation changed.
+
+        Check relation data against relation data inserted in the database
+        from a read-only endpoint.
+        """
+        if self.unit.is_leader():
+            # run from a non leader unit
+            # to force consume data from relation databag
+            return
+
+        if "inserted" not in self._peers.data[self.app]:
+            # run only after flag is set
+            event.defer()
+            return
+
+        # get remote relation databag
+        remote_relation = self.model.get_relation(REMOTE)
+        if not remote_relation:
+            event.defer()
+            return
+
+        remote_data = remote_relation.data[remote_relation.app]
+
+        # parse read-only database host
+        ro_database = remote_data["read-only-endpoints"].split(",")[0].split(":")[0]
+
+        config = {
+            "user": remote_data["username"],
+            "password": remote_data["password"],
+            "host": ro_database,
+            "database": self.database_name,
+            "raise_on_warnings": False,
+        }
+
+        with MysqlConnector(config, commit=False) as cursor:
+            rows = self._read_test_data(cursor, remote_relation.id)
```

Comment:
Probably it is a good idea to also test that it is really a read only endpoint by trying to write some data to it. Feel free to dismiss this idea (I know that it is being tested in some way in https://github.com/canonical/mysql-k8s-operator/blob/bc5d536d4bd3630e16386dcb440652b670637b00/tests/unit/test_database.py#L35)

Reply:
that's very reasonable